xarray/core/dataset.py

```diff
@@ -1485,7 +1485,7 @@ def __getitem__(self, key):
         if hashable(key):
             return self._construct_dataarray(key)
         else:
-            return self._copy_listed(np.asarray(key))
+            return self._copy_listed(np.asarray(key, dtype="O"))
```

Comment:
Why is `np.asarray` needed here?
Can we just check that `key` is a list, and raise an error otherwise.

Right now, if `key` is for example a set (i.e., `ds[set(["foo"])]`), an error is raised anyways because `np.asarray` creates a 0-d array and iteration is not possible. The error is not raised directly by `xarray` though and I think it's not very clear, especially because the original set is iterable.

Also, is the goal to only allow lists? Does it make sense to allow any iterable?

Reply:
https://github.com/pydata/xarray/blob/5d367db52a9044d2eeec5d460a9ea2f2f69d10e4/xarray/core/dataset.py#L1345
Seems to only be used to loop through in `._copy_listed`. I can't really think of a reason why it has to be forced to ndarray? names is also typed as an Iterable[Hashable], so I suppose it makes sense to allow any iterable.

Reply:
Does anything break if we remove it?

Reply:
All tests look OK removing `np.asarray`

Reply:
Great! Without wanting to ignore Chesterton's fence — let's do it unless we hear objections in the next day or so.